### PR TITLE
Add/36019 php product form helper classes

### LINKED ISSUE #36019: [Product Extensibility] Create PHP helper function for adding new fields
**Description**
This is part of #35953 epic.
We want to give plugin developers the ability to extend the new product management screen by use of PHP.
This issue will be the very first for laying the ground work and adding this support for fields specifically.

**NOTE: We will support templates in the future, so be aware of how this may impact the work.**

We should expose a public PHP function that can be triggered within a `init` action, something like `wc_add_product_form_field` (or tag `admin` in there to make it more admin specific. But similar to either the [`add_settings_field`](https://developer.wordpress.org/reference/functions/add_settings_field/) or [`add_meta_box`](https://developer.wordpress.org/reference/functions/add_meta_box/) functions. We may want to base the implementation of these two examples.

**Acceptance criteria**
- [ ] `add_product_form_field` should have an `id` and ask for a plugin `id` (will be used later)
- [ ] `add_product_form_field` should provide a location prop for where it should be displayed (should match the location prop within a SlotFill see: https://github.com/woocommerce/woocommerce/issues/36014 )
- [ ] `add_product_form_field` should provide `args` for the input field, we may want to expose `type`, `name`, `label` for now (and others if you think they are important) we can always expand on this.
- [ ] Field data should be accessible within data (maybe a global constant? that is what `add_settings_field` uses)

### PR DESCRIPTION
**Pull request by louwie17**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adds some of the initial ProductForm classes for 3PD's to extend the new Product form through PHP.
This currently only adds the ability to add fields.

Closes #36019

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

### How to test the changes in this Pull Request:

1. Run PHP unit tests locally to see if they all pass, see https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/tests/README.md#initial-setup for initial setup and how to.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/src/Internal/Admin/ProductForm/ComponentTrait.php
Comment (joshuatf) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/ComponentTrait.php:26]: Cool!  Just out of curiosity, do we have plans as for how this might be used?  I assume the "slug/id" will be self-reported by the plugin adding it, right?

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php
Comment (joshuatf) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php]: Will there be a separate `VariationForm` class for variations?  Just thinking out loud for the sake of this logic and reducing ambiguity in this class name if necessary.

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php
Comment (joshuatf) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php]: Should this return the newly created instance of `Field` or a `WP_Error` if arguments are missing?

It might be useful to consumers and would make testing slightly easier.

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php
Comment (joshuatf) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php]: ©️ 🍝

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php
Comment (joshuatf) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php]: According to https://github.com/woocommerce/woocommerce/issues/36019 we should also include a `label` property, right?

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Field.php
Comment (joshuatf) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/Field.php]:
```suggestion
 * Field class.
```

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Field.php
Comment (joshuatf) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/Field.php]:
```suggestion
	 * Component traits.
```

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Field.php
Comment (joshuatf) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/Field.php]:
```suggestion
	 * Array of required arguments.
```

File: plugins/woocommerce/src/Internal/Admin/ProductForm/ComponentTrait.php
Comment (joshuatf) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/ComponentTrait.php]:
```suggestion
 * ComponentTrait class.
```

**Review by joshuatf — COMMENTED** [review]
Great work overall, Lourens!  This looks like a good approach to me.  Left a few clean-up comments below.

I'm wondering if we should also have a `Section` class, but this can (and should be) part of a follow-up if we decide to do it.

File: plugins/woocommerce/src/Internal/Admin/ProductForm/ComponentTrait.php
Comment (louwie17) [reply plugins/woocommerce/src/Internal/Admin/ProductForm/ComponentTrait.php:26]: > do we have plans as for how this might be used?

No not yet, I assume we might want to have an alternate function to register a plugin that may include an `icon` and a full name. Or we did want to force this to the actual `plugin` slug and check with the plugin list.
We haven't fully thought this through yet, but maybe matching this with the `plugin` slug for now would be a smart plan.

> I assume the "slug/id" will be self-reported by the plugin adding it, right?

Yeah, as there is no way to tell otherwise what plugin registered a field. So technically they could reference `woocommerce` as well, unless we prevent that from being allowed.

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php
Comment (louwie17) [reply plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php]: My thought was to actually only keep `Form` and use that class to register new templates, tabs, sections, and fields. I didn't really want to make a standard were we create a new `Form` class for every type of `Form`.

Sounds like Variations might actually be the only exception to this approach which we may of not fully thought through yet. 
I guess I have two thoughts around this:
- Do we want to have the Variation form be defined as a custom template, not available for selection (product creation) but only used to render the form layout for Variations?
- Or do we want to create a new `VariationForm` that mostly inherits from `Form` but is for Variations only.

I would lean towards the first approach more, as we want to treat variations largely the same.

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php
Comment (louwie17) [reply plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php]: Yeah that's a good point, I will update that.

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php
Comment (louwie17) [reply plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php]: You got it :)

**Comment by louwie17** [comment]
Thanks for the review @joshuatf 

> I'm wondering if we should also have a `Section` class, but this can (and should be) part of a follow-up if we decide to do it.

Yeah the plan was to do this as a follow up, although I end up adding it in here as well, including a `Card` class as well.

**Comment by louwie17** [comment]
@joshuatf  this should be good for a re-review if you have a chance.
I did try to consolidate some of the logic around the Field, Card, and Section classes as they overlap a lot. Let me know what you think or if we should take a different approach.

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Component.php
Comment (joshuatf) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/Component.php]: Nitpicky, but this name sounds like it would retrieve all arguments passed to the constructor.  Maybe this could be more specific like `get_form_arguments` or `get_additional_args`.

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php
Comment (joshuatf) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php]: The term `card` is tied to a UI element, which forces us to continue using this UI element or break the API in the future if the design changes.

Should we update this to something more generic like `add_subsection` or `add_group`?  We might need to agree on these terms which may be p2 worthy, but I'd take whatever inspiration we can from the legacy experience.

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php
Comment (joshuatf) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php]: Should we check to make sure this class exists before using it just to be safe?

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php
Comment (joshuatf) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php]: This abstraction took me a while to follow it backwards.  I'm wondering if it might just be easier to do something like:

```
try {
  new $class( $args );
} catch {
  return new WP_Error(...);
}
```

And then handle any exceptions inside the class itself instead of the parent.

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Field.php
Comment (joshuatf) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/Field.php:41]: Just curious why we nest other arguments inside of properties.  Is this something we did in the previous API that we're following?

**Review by joshuatf — COMMENTED** [review]
Left a few more comments, @louwie17.  Overall this looks pretty good!  Just want to clarify a few things before merging.

Also, I think it would be good to get some p2 discussion around different patterns with abstractions/factories (does not need to block this PR) to see if other teams have run into blockers with different approaches and have experience to share.

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Field.php
Comment (louwie17) [reply plugins/woocommerce/src/Internal/Admin/ProductForm/Field.php:41]: We do something similar with the [`add_settings_fields`](https://developer.wordpress.org/reference/functions/add_settings_field/) with the `args` param.

I mostly wanted to differentiate what arguments would be specifically for the html field used as a property and which are more for general details like order/section/type etc....
But instead of relying on param order, I thought it was easier to pass them in as an argument array.

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php
Comment (louwie17) [reply plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php]: Good suggestion, I made that change!

**Comment by louwie17** [comment]
@joshuatf this should be good for a re-review. I also end up renaming the `Form` class to `FormFactory` to keep it more clear.

File: plugins/woocommerce/src/Internal/Admin/ProductForm/README.md
Comment (joelclimbsthings) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/README.md:11]: I believe this line should be `\FormFactory` as well, instead of `\Form`

File: plugins/woocommerce/src/Internal/Admin/ProductForm/Component.php
Comment (joelclimbsthings) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/Component.php:9]: You have a few comments around here that still reference this as the _Card_ class instead of _Component_

File: plugins/woocommerce/src/Internal/Admin/ProductForm/FormFactory.php
Comment (joelclimbsthings) [inline plugins/woocommerce/src/Internal/Admin/ProductForm/FormFactory.php:206]: Totally optional, but this looks like it could be as tid bit fragile if there was ever a name change/etc here. Could do a mapping or something similar, but not required at all.

**Review by joelclimbsthings — APPROVED** [review]
Great job @louwie17 . I left a few very minor comments, but looks pretty well ready to go to me. Going to pre-approve, and you can choose to address those comments in another PR if you want to get this merged since it's blocking.

**Comment by louwie17** [comment]
> Great job @louwie17 . I left a few very minor comments, but looks pretty well ready to go to me. Going to pre-approve, and you can choose to address those comments in another PR if you want to get this merged since it's blocking.

Thanks @joelclimbsthings for the sake of moving some unblocked work forward I will address these changes in the API PR -> https://github.com/woocommerce/woocommerce/pull/36165.
I appreciate the review.

**Pull request merged**
